### PR TITLE
Add App Pool metrics to IIS module

### PR DESCRIPTION
**Pull request by simonwillcock**
As discussed in #69 this adds metrics exposed by the `Win32_PerfRawData_APPPOOLCountersProvider_APPPOOLWAS` class

### CODE REVIEW

**Comment by carlpett** [comment]
Nice, really good work! This looks pretty good from a quick lookover. The one thing I would like to change is to map the `current_application_pool_state` to readable labels instead of a number. The recommended pattern is to have a gauge with a label for each state and the value 1 for the currently active state, so you'd get something like this:

```
current_application_pool_state{app="foo",state="uninitialized"} 0
current_application_pool_state{app="foo",state="initialized"} 0
current_application_pool_state{app="foo",state="running"} 1
...
```

So basically, add another label to `CurrentApplicationPoolState`, then do a bit of mapping before sending it down the channel.

**Comment by simonwillcock** [comment]
Ah good thinking. I'll see if I can work out how to do that.

**Comment by simonwillcock** [comment]
How does that look? I wasn't sure if the state values should be lower case and/or snake case? I've left them proper cased with spaces as they were described in the docs for now.

**Comment by carlpett** [comment]
Nice! Only one minor point, we should expose even the states which are not currently active, but with a value of 0. I think we're doing this somewhere already, let's see if I can find it.

**Comment by simonwillcock** [comment]
Ah beautiful - thanks for that. I'll make that change, and so should the value of the metric be just 1 for the active states or leave them as their current value? Having them as just 1 might make them more easier to use when querying perhaps?

**Comment by carlpett** [comment]
Exactly, set it to 1 if it is the active state, 0 otherwise. Makes for easy querying, as you say.

**Comment by simonwillcock** [comment]
I think that should do the trick. Thanks for the example too, that helped a lot 😄

**Comment by martinlindhe** [comment]
Looks fantastic! Great job @simonwillcock

File: collector/iis.go [hunk 1]
```diff
@@ -650,13 +650,20 @@ func (c *IISCollector) collect(ch chan<- prometheus.Metric) (*prometheus.Desc, e
...
-		)
+		for key, label := range ApplicationStates {
+			isCurrentState := 0.0
+			if key == app.CurrentApplicationPoolState {
+				isCurrentState = 1.0
+			}
+			ch <- prometheus.MustNewConstMetric(
+				c.CurrentApplicationPoolState,
+				prometheus.GaugeValue,
+				float64(isCurrentState),
```
Comment (martinlindhe) [inline collector/iis.go]: I think this cast is unneeded, as you create a float value implicit already by `isCurrentState := 0.0`

Comment (simonwillcock) [inline collector/iis.go]: Ah, right you are. I'll fix that up

**Comment by martinlindhe** [comment]
Looks all good to me, but since @carlpett started the review, I'll let him merge it

**Comment by carlpett** [comment]
Took it for a test run on one of my machines, and I think there might be something off with the uptime-metrics:
`wmi_iis_current_application_pool_uptime{app=".NET v4.5 Classic"} 1.3129207244518765e+17`

In the [service collector](https://github.com/martinlindhe/wmi_exporter/blob/master/collector/system.go#L122) we do some normalization to get around this, but it doesn't look like that exact normalization works here (I get 17000 days of uptime). Trying to find some documentation.

**Comment by martinlindhe** [comment]
@carlpett days since Jan 1, 1970 = 17277, related?

**Comment by carlpett** [comment]
Aha, well spotted. The `Timestamp_Object` property contains the current timestamp, so `(Timestamp_Object - TotalApplicationPoolUptime)/Frequency_Object` should give the uptime in seconds, I think?

**Comment by simonwillcock** [comment]
Oh good pickup!

**Comment by brian-brazil** [comment]
If it's available, expose the time it started rather than uptime. This lets you use `changes()` to count restarts.

**Comment by simonwillcock** [comment]
I don't believe it is unfortunately - only uptime

**Comment by simonwillcock** [comment]
Quick question - do I need to cast the result of that equation? Or cast the value first and then do the calculation? 

Something like this `float64((Timestamp_Object - app.CurrentApplicationPoolUptime) / Frequency_Object),` ?

**Comment by simonwillcock** [comment]
Ended up doing the same what is done in `system.go` and casting each property separately. Hope that is what we're going for :)

**Comment by martinlindhe** [comment]
@simonwillcock if the underlying type of a and b is an int, then float64(a / b) will first calc the integer result of a/b, then cast to float, so you'll loose precision. The way you solved it is correct!

**Comment by simonwillcock** [comment]
Ah rightio, thanks for explaining that for me!

**Comment by brian-brazil** [comment]
> I don't believe it is unfortunately - only uptime

Are you sure CurrentApplicationPoolUptime isn't the time it started at?

That you're doing a subtraction indicates to me that that number is not uptime.

**Comment by simonwillcock** [comment]
Its possible that I'm misunderstanding.. the docs I found said this for CurrentApplicationPoolUptime: "The length of time, in seconds, that the application pool has been running since it was started" but they didn't appear to be official msdn docs though.

**Comment by carlpett** [comment]
@simonwillcock Yeah, the docs are for the processed metrics that Windows uses in ProcMon, so you'll have to take those with a pinch of salt...

**Comment by simonwillcock** [comment]
Interesting. I think it sounds like you guys know about this than me so I'll take your word for it! 
Once I'm back at work after the weekend I'll check the value it's returning and can try and give a definitive answer if we don't have before then.

**Comment by simonwillcock** [comment]
It looks like @brian-brazil is on the money and `CurrentApplicationPoolUptime` is actually the start time! This is actually the same type of value as `SystemUpTime` in `system.go`. 

I'll push an update to make it return the uptime instead so that it matches `system.go` for consistency. 
Good pickup!

**Comment by carlpett** [comment]
@simonwillcock Great! Just one last thing (sorry): with this change, the metric should probably be renamed from `_uptime` to `_start_time`? Apart from that, I'm good to merge!

**Comment by carlpett** [comment]
@simonwillcock Thanks, and kudos on a very nice first Go PR!

**Comment by simonwillcock** [comment]
Thanks very much for guiding me through it 😊

**Pull request merged**